src/context.rs

```diff
@@ -158,6 +162,57 @@ impl Context {
         self.blobdir.as_path()
     }
 
+    pub fn copy_to_blobdir(&self, orig_filename: impl AsRef<str>) -> Result<String> {
+        // return a $BLOBDIR/<filename> with the content of orig_filename
+        // copied into it. The <filename> will be safely derived from
+        // orig_filename, and will not clash with existing filenames.
+        let dest = self.new_blob_file(&orig_filename, b"")?;
+        if dc_copy_file(
+            &self,
+            PathBuf::from(orig_filename.as_ref()),
+            PathBuf::from(&dest),
+        ) {
+            Ok(dest)
+        } else {
+            bail!("could not copy {} to {}", orig_filename.as_ref(), dest);
+        }
+    }
+
+    pub fn new_blob_file(&self, orig_filename: impl AsRef<str>, data: &[u8]) -> Result<String> {
+        // return a $BLOBDIR/<FILENAME> string which corresponds to the
+        // respective file in the blobdir, and which contains the data.
+        // FILENAME is computed by looking and possibly mangling the
+        // basename of orig_filename. The resulting filenames are meant
+        // to be human-readable.
+        let (stem, ext) = dc_derive_safe_stem_ext(orig_filename.as_ref());
+
+        // ext starts with "." or is empty string, so we can always resconstruct
+
+        for i in 0..3 {
+            let candidate_basename = match i {
+                // first a try to just use the (possibly mangled) original basename
+                0 => format!("{}{}", stem, ext),
+
+                // otherwise extend stem with random numbers
+                _ => {
+                    let mut rng = thread_rng();
+                    let random_id: u32 = rng.gen();
```

Comment:
`let random_id: u32 = rand::random();` can do these two lines in one simpler one afaik.

Reply:
hum, i took what i found elsewhere.  Are you sure it's the same?

Reply:
it's what the docs say ;)